File: omnivault-backend/src/main/java/com/personal/omnivault/service/ContentService.java
package com.personal.omnivault.service;

import com.personal.omnivault.domain.dto.request.ContentUpdateRequest;
import com.personal.omnivault.domain.dto.request.LinkContentCreateRequest;
import com.personal.omnivault.domain.dto.request.TextContentCreateRequest;
import com.personal.omnivault.domain.dto.response.ContentDTO;
import com.personal.omnivault.domain.model.Content;
import com.personal.omnivault.domain.model.ContentType;
import org.springframework.core.io.Resource;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
import java.util.UUID;

public interface ContentService {

    /**
     * Get content by ID
     *
     * @param contentId The content ID
     * @return The content DTO
     */
    ContentDTO getContent(UUID contentId);

    /**
     * Get content entity by ID
     *
     * @param contentId The content ID
     * @return The content entity
     */
    Content getContentEntity(UUID contentId);

    /**
     * Get all content with pagination
     *
     * @param pageable Pagination information
     * @return Page of content
     */
    Page<ContentDTO> getAllContent(Pageable pageable);

    /**
     * Get content by folder
     *
     * @param folderId The folder ID
     * @param pageable Pagination information
     * @return Page of content
     */
    Page<ContentDTO> getContentByFolder(UUID folderId, Pageable pageable);

    /**
     * Get content by type
     *
     * @param contentType The content type
     * @param pageable Pagination information
     * @return Page of content
     */
    Page<ContentDTO> getContentByType(ContentType contentType, Pageable pageable);

    /**
     * Get content by tag
     *
     * @param tagId The tag ID
     * @param pageable Pagination information
     * @return Page of content
     */
    Page<ContentDTO> getContentByTag(UUID tagId, Pageable pageable);

    /**
     * Get favorite content
     *
     * @param pageable Pagination information
     * @return Page of content
     */
    Page<ContentDTO> getFavoriteContent(Pageable pageable);

    /**
     * Get recent content
     *
     * @param pageable Pagination information
     * @return Page of content
     */
    Page<ContentDTO> getRecentContent(Pageable pageable);

    /**
     * Get popular content
     *
     * @return List of popular content
     */
    List<ContentDTO> getPopularContent();

    /**
     * Create text content
     *
     * @param request The text content create request
     * @return The created content DTO
     */
    ContentDTO createTextContent(TextContentCreateRequest request);

    /**
     * Create link content
     *
     * @param request The link content create request
     * @return The created content DTO
     */
    ContentDTO createLinkContent(LinkContentCreateRequest request);

    /**
     * Create file content
     *
     * @param file The file to upload
     * @param title The content title
     * @param description The content description
     * @param folderId The folder ID
     * @param tagIds List of tag IDs
     * @return The created content DTO
     */
    ContentDTO createFileContent(
            MultipartFile file,
            String title,
            String description,
            UUID folderId,
            List<UUID> tagIds,
            List<String> newTags
    );

    /**
     * Update content
     *
     * @param contentId The content ID
     * @param request The content update request
     * @return The updated content DTO
     */
    ContentDTO updateContent(UUID contentId, ContentUpdateRequest request);

    /**
     * Toggle favorite status
     *
     * @param contentId The content ID
     * @return The updated content DTO
     */
    ContentDTO toggleFavorite(UUID contentId);

    /**
     * Update content tags
     *
     * @param contentId The content ID
     * @param tagIds List of tag IDs
     * @param newTags List of new tag names
     * @return The updated content DTO
     */
    ContentDTO updateContentTags(UUID contentId, List<UUID> tagIds, List<String> newTags);

    /**
     * Delete content
     *
     * @param contentId The content ID
     */
    void deleteContent(UUID contentId);

    /**
     * Get content file as resource
     *
     * @param contentId The content ID
     * @return The file resource
     */
    Resource getContentFile(UUID contentId);

    /**
     * Get content thumbnail as resource
     *
     * @param contentId The content ID
     * @return The thumbnail resource
     */
    Resource getContentThumbnail(UUID contentId);

    /**
     * Search content
     *
     * @param searchTerm The search term
     * @param pageable Pagination information
     * @return Page of content
     */
    Page<ContentDTO> searchContent(String searchTerm, Pageable pageable);

    /**
     * Increment view count
     *
     * @param contentId The content ID
     */
    void incrementViewCount(UUID contentId);
}

File: omnivault-frontend/src/services/contentService.js
import axiosInstance from "./axiosInstance";
import { apiCache } from "../utils/apiCache";
import fileService from "./fileService";

const contentService = {
  getAllContent: async (page = 0, size = 10) => {
    const response = await axiosInstance.get(
      `/contents?page=${page}&size=${size}&sort=createdAt,desc`
    );
    return response.data;
  },

  getContent: async (contentId) => {
    const cacheKey = `content_${contentId}`;

    // Try to get from cache first and check if it's fresh (less than 5 minutes old)
    const cachedContent = apiCache.get(cacheKey);
    if (cachedContent && cachedContent._timestamp) {
      const now = Date.now();
      const ageInMinutes = (now - cachedContent._timestamp) / (1000 * 60);

      // Use cache if it's less than 5 minutes old
      if (ageInMinutes < 5) {
        return cachedContent;
      }
    }

    // If not in cache or stale, fetch from API
    const response = await axiosInstance.get(`/contents/${contentId}`);

    // Add timestamp to the response data
    const responseWithTimestamp = {
      ...response.data,
      _timestamp: Date.now(),
    };

    // Store in cache before returning
    apiCache.set(cacheKey, responseWithTimestamp);
    return responseWithTimestamp;
  },

  getContentByType: async (contentType, page = 0, size = 10) => {
    const response = await axiosInstance.get(
      `/contents/type/${contentType}?page=${page}&size=${size}`
    );
    return response.data;
  },

  getContentByTag: async (tagId, page = 0, size = 10) => {
    const response = await axiosInstance.get(
      `/contents/tag/${tagId}?page=${page}&size=${size}`
    );
    return response.data;
  },

  getFavorites: async (page = 0, size = 10) => {
    const response = await axiosInstance.get(
      `/contents/favorites?page=${page}&size=${size}`
    );
    return response.data;
  },

  getRecent: async (page = 0, size = 10) => {
    const response = await axiosInstance.get(
      `/contents/recent?page=${page}&size=${size}`
    );
    return response.data;
  },

  getPopular: async () => {
    const response = await axiosInstance.get("/contents/popular");
    return response.data;
  },

  createTextContent: async (textContentData) => {
    const response = await axiosInstance.post(
      "/contents/text",
      textContentData
    );
    return response.data;
  },

  createLinkContent: async (linkContentData) => {
    const response = await axiosInstance.post(
      "/contents/link",
      linkContentData
    );
    return response.data;
  },
  getFolderContent: async (folderId, page = 0, size = 10) => {
    const response = await axiosInstance.get(
      `/contents/folder/${folderId}?page=${page}&size=${size}`
    );
    return response.data;
  },

  uploadFile: async (
    file,
    title,
    description,
    folderId,
    tagIds,
    newTags,
    storageLocation
  ) => {
    const formData = new FormData();
    formData.append("file", file);

    if (title) formData.append("title", title);
    if (description) formData.append("description", description);
    if (folderId) formData.append("folderId", folderId);
    if (storageLocation) formData.append("storageLocation", storageLocation);

    if (tagIds && tagIds.length > 0) {
      tagIds.forEach((tagId) => formData.append("tagIds", tagId));
    }

    if (newTags && newTags.length > 0) {
      newTags.forEach((tag) => formData.append("newTags", tag));
    }

    const response = await axiosInstance.post("/contents/file", formData, {
      headers: {
        "Content-Type": "multipart/form-data",
      },
    });

    return response.data;
  },

  updateContent: async (contentId, contentData) => {
    // Invalidate cache for this content
    apiCache.remove(`content_${contentId}`);
    // Also clear any thumbnail caches that might include this content
    apiCache.clear(`thumbnails_`);

    const response = await axiosInstance.put(
      `/contents/${contentId}`,
      contentData
    );
    return response.data;
  },

  toggleFavorite: async (contentId) => {
    const response = await axiosInstance.put(`/contents/${contentId}/favorite`);
    return response.data;
  },

  deleteContent: async (contentId) => {
    await axiosInstance.delete(`/contents/${contentId}`);
  },

  searchContent: async (query, page = 0, size = 10) => {
    const response = await axiosInstance.get(
      `/contents/search?query=${query}&page=${page}&size=${size}`
    );
    return response.data;
  },

  getFileUrl: (contentId) => {
    return fileService.getFileUrl(contentId);
  },

  getThumbnailUrl: (contentId) => {
    return fileService.getThumbnailUrl(contentId);
  },

  fetchBatchThumbnails: async (contentIds) => {
    return fileService.fetchBatchThumbnails(contentIds);
  },

  // These methods remain the same for programmatic access
  fetchFileWithAuth: async (contentId) => {
    return fileService.fetchFileWithAuth(contentId);
  },

  fetchThumbnailWithAuth: async (contentId) => {
    return fileService.fetchThumbnailWithAuth(contentId);
  },

  // Get a presigned URL for cloud content
  getContentPresignedUrl: async (contentId) => {
    const response = await axiosInstance.get(
      `/contents/${contentId}/cloud-url`
    );
    return response.data.url;
  },

  // Get a presigned URL for a thumbnail
  getThumbnailPresignedUrl: async (contentId) => {
    const response = await axiosInstance.get(
      `/contents/${contentId}/thumbnail-url`
    );
    return response.data.url;
  },

  // Move content between storage locations
  moveContentStorage: async (contentId, targetLocation) => {
    const response = await axiosInstance.put(
      `/contents/${contentId}/storage?targetLocation=${targetLocation}`
    );
    return response.data;
  },

  downloadFile: async (contentId, filename) => {
    try {
      // First check if it's cloud storage with a presigned URL
      try {
        const presignedUrl = await contentService.getContentPresignedUrl(
          contentId
        );
        if (presignedUrl) {
          // Open in a new window/tab or trigger download via window.open
          window.open(presignedUrl, "_blank");
          return;
        }
      } catch (error) {
        // If not cloud or error getting URL, fall back to direct download
        console.log(
          "Not cloud storage or error getting presigned URL, falling back to direct download",
          error
        );
      }

      // Original local storage download logic
      const response = await axiosInstance.get(`/contents/${contentId}/file`, {
        responseType: "blob",
      });

      // Create a blob URL and trigger download
      const url = window.URL.createObjectURL(new Blob([response.data]));
      const link = document.createElement("a");
      link.href = url;
      link.setAttribute("download", filename || "download");
      document.body.appendChild(link);
      link.click();
      link.remove();

      // Clean up
      setTimeout(() => {
        window.URL.revokeObjectURL(url);
      }, 100);
    } catch (error) {
      console.error("Error downloading file:", error);
    }
  },
};

export default contentService;
